### ISSUE TITLE
New windows cover dock when in intellihide mode.

### ISSUE THREAD

**Issue by jovaha** [micheleg/dash-to-dock#105]
When in intellihide mode many gnome applications start on the dock which makes it hide. It would be preferable if new windows avoided the dock like when the dock is always visible but still kept the features of intellihide.

**Comment by ermo** [comment]
The issue in the OP is the only issue I have with the dash to dock intellihide feature (which is otherwise awesome!).  My suggestion ties into the OP like such:

* If there are no visible windows on a workspace, it's nice to have the dock show up.
* If I open a new app on an otherwise empty workspace, it's nice to have the dock show (which is what the OP wants as well I think).
* If I've gone full screen in an app on the current workspace (no matter whether the full-screen app is in the foreground or in the background), it's nice to NOT have the dock show unless I push the edge (otherwise it will be obscured when I switch to another app or anchor an app to the right of the workspace).
* If I've anchored an app to the left of the workspace (no matter whether it's in the foreground or in the background) , it's nice to NOT have the dock show unless I push the edge (otherwise it will be obscured when I switch to another app or anchor an app to the right of the workspace).

So, assuming that the dock is anchored to the left side of the monitor (i.e. the default), I'm envisioning a method where the dock only (intelli)hides when the user either:

1) Maximizes a window 
or
2) Anchors a window to the left side of the monitor with Win+left

Am I making sense?

**Comment by micheleg** [comment]
If I have understood it correctly, you are defining the "intelligent autohide" as (relative to the current behaviour): "the dock is fixed and always visible. It hides when a windows get maximized or anchored to half-screen (possibly only on the side of the dock), and not only with fullscreen windows as now ". This would mean not having the dodging features at all. 

Or equivalently, you want an option for the intellihide disabled case where the dock hides on situation 1) and 2). Did I understand correctly?

**Comment by ermo** [comment]
It sounds like we're on the same page, yes.  I guess you could call my proposal 'limited intellihide'.

As it stands (and as the OP points out), the dodging behaviour is -- ironically enough -- 'in the way' in a sense.  I'd love to try my proposal out in practice, but since you'd probably be the one actually implementing it (knowing the code intimately and all), it is obviously totally your call whether you find it worth trying out. =)

**Comment by micheleg** [comment]
It should not be too difficult to hack this feature for a quick test so that we can see if it actually works in practice. One of these days I'll try to play with it.

**Comment by ermo** [comment]
Sounds cool! :+1:

**Comment by micheleg** [comment]
@ermo, I did a quick test: unfortunately the result of tweaking the fixed mode is ugly, because I can let the window expand  to the full sceen area only after I get notified it is maximized, but it is too late to have a smooth transition. You always get and ugly jump of it. As far as I know I can't do an operation before the maximization of the window happens. I realized I did try the same thing at a certain point, with similar results.

An alternative approach is instead in the branch experimental_intellihide_only_maximized which in intellihide modes just considers all maximized windows . This is similar to  what you were asking, but then you  come against the same problem explained above, i.e. the inability of tell mutter to not put newly open  windows above the dock, which is the origin of the problem. I can move them just after they are created, but again you get an ugly jump.

If you are aware of any extension which claim it can move windows, change the animation,  or related please let me know.

**Comment by micheleg** [comment]
It turned out there's a at least in GNOME Shell 3.14 an option for mutter to force new windows to be in the center of the screen ( /org/gnome/mutter/center-new-windows ). It's not optimal but better than nothing.

**Comment by micheleg** [comment]
@ermo, @jovaha I'm getting there. The last commit in the same branch should force newly opened windows to stay away from the dock ( leaving a 5px gap ). I noticed that the algorithm putting windows gets sometimes confused by this, but only in specific edge cases. I would be glad if you could test it and report any problem. The branch is for 3.14, but I can easily rebase it into 3.16 if you want.

**Comment by ermo** [comment]
@micheleg: Really appreciate the time you are putting into prototyping this.  You're essentially rewriting the interaction pattern with GNOME 3 to be something else than what the GNOME 3 devs envisioned, so there's bound to be some tilting at windmills when going down this path.

I'll have a look as I'm on GNOME 3 right now.  Do you have any tips on how to easily switch between extension versions/branches?

**Comment by micheleg** [comment]
If you want to test the development version  - I assume you have basic knowledge of git - of the extension, clone locally the repository, checkout the branch you are interested in  - you need to be sure the branch is for your shell version - and run ```make install```. Afterwards, restart the shell with ```alt+f2 , r, ENTER```, or run ```gnome-shell --replace``` in a terminal (this will show eventual errors in the terminal, but remember that closing the terminal will terminate the shell too).

I have to say that I really like this settings with the dock autohiding with maximized windows and always visible otherwise. I hope the trick to move the windows doesn't turn out to bring problems.

**Comment by ermo** [comment]
Ok, just tested the development version.  Compared to how it works now, I'll try to describe what I was hoping for.

These are the settings I'm using:

![dash_to_dock_1](https://cloud.githubusercontent.com/assets/118045/7798782/ae6461b2-02fe-11e5-8977-7cde16518663.png)
![dash_to_dock_2](https://cloud.githubusercontent.com/assets/118045/7798781/ae622ed8-02fe-11e5-9519-e37e04e5f612.png)
![dash_to_dock_3](https://cloud.githubusercontent.com/assets/118045/7798780/ae61ed7e-02fe-11e5-8efd-187c23b47779.png)
![dash_to_dock_4](https://cloud.githubusercontent.com/assets/118045/7798779/ae5ee598-02fe-11e5-9d77-53f299757585.png)


**Expected behaviour:**

1. When I move a window to the left of the desktop (dragging the window by its title bar) and encounter the edge of the dock, I would expect the window to snap to the edge of the dock, not go under it.
2. If I keep dragging the window to the left edge of the screen, the effect should be the same as Win+Left

**Observed behaviour:**

1. The window goes under the dock -- it doesn't snap to the dock
2. as described in Expected behaviour.

![dash_to_dock_5](https://cloud.githubusercontent.com/assets/118045/7798853/b3ef5492-02ff-11e5-8aea-9b2c15e70c67.png)

**Comment by micheleg** [comment]
Mmmm, what your are missing is just the snapping of windows on the dock edge. I assume you mean the effect that stops the windows movement unless you put more force. That, as far as I know is managed by mutter, rather then gnome shell, and my previous attempts to indirectly control it have not been successful. I can re-look into it, but I'm pessimistic.

**Comment by ermo** [comment]
For the sake of clarity, let me try to describe the different states for my proposed version of "intelligent autohide":

1) Dock is visible and maximized on the left with the app-icon at the top by default, using your custom theme. No windows are visible on the desktop.

![dash_to_dock_empty_desktop](https://cloud.githubusercontent.com/assets/118045/7814189/c15801c6-03c0-11e5-83eb-538a13a7e331.png)

2) When opening an app, its window will never obscure the dock -- if necessary, it will be resized to fit the desktop area to the right of the dock.

![dash_to_dock_app_not_snapped_to_dock_and_not_maximized](https://cloud.githubusercontent.com/assets/118045/7814104/28535c82-03c0-11e5-8260-bf699fa27753.png)

3) If a window is dragged from e.g. the middle of the desktop to the desktop-facing (i.e. right) edge of the dock via its titlebar, it will be stopped by the outer edge of the dock, just as if there was no dodge behaviour. Think of this as 'snap-to-dock'.

![dash_to_dock_app_snapped_to_dock_but_not_maximized_nor_pinned_left](https://cloud.githubusercontent.com/assets/118045/7814121/431ad5b8-03c0-11e5-8a9b-5bfd162f7d08.png)

4) If I drag a currently 'snapped-to-dock' window to the left across the area occupied by the dock  such that the 'grabbing hand' icon moves to (or past) the very edge of the left side of the monitor edge, the window will be pinned to the left edge like if I had used Win+Left.  Successfully completing a 'pin-to-edge' operation will make the dock autohide and the dock will only be shown again if I move the pointer to the left side of the monitor in a manner similar to the current 'push to show' autohide behaviour.  If I abandon the drag operation before the 'grabbing hand' icon moves to (or past) the very edge of the left side of the monitor, the window will revert to its previous 'snap-to-dock' state from 3).  The dock is fully visible until the drag operation is successfully completed.

![dash_to_dock_app_is_pinned_left_and_push_to_show_autohide_is_active](https://cloud.githubusercontent.com/assets/118045/7814166/9fe8af40-03c0-11e5-9901-dad979203838.png)

4a) If a window is maximized, use the same push-to-show autohide behavior as 4).

![dash_to_dock_app_is_maximized_and_push_to_show_autohide_is_active](https://cloud.githubusercontent.com/assets/118045/7814169/a65f34c0-03c0-11e5-9593-71b17d9a9054.png)


I hope the above is sufficiently clear to you. Thank you for reading. :)

**Comment by ermo** [comment]
I don't yet have GNOME 3.18, but reading your release notes for v49 it seems that you have actually succeeded in implementing my above feature request?

:+1 :yum:

**Comment by micheleg** [comment]
Nope, I just added an option to hide the dock only with maximized windows. Unfortunately, all the rest is the same, so windows still tend to cover the dock when they are first created and windows snapping happens to the screen border only. However, this was the first step in this direction. Next is opening new windows in the right place.
